evalml/pipelines/components/transformers/imputers/knn_imputer.py

```diff
@@ -0,0 +1,128 @@
+        random_seed (int): Seed for the random number generator. Defaults to 0.
+
+    """
+
+    name = "KNN Imputer"
+
+    def __init__(self, number_neighbors=3, random_seed=0, **kwargs):
+        parameters = {"number_neighbors": number_neighbors}
+        parameters.update(kwargs)
+
+        imputer = Sk_KNNImputer(
+            n_neighbors=number_neighbors,
+            missing_values=np.nan,
+            **kwargs,
+        )
+        self._all_null_cols = None
+        super().__init__(
+            parameters=parameters,
+            component_obj=imputer,
+            random_seed=random_seed,
+        )
+
+    def fit(self, X, y=None):
+        """Fits imputer to data. 'None' values are converted to np.nan before imputation and are treated as the same.
+
+        Args:
+            X (pd.DataFrame or np.ndarray): the input training data of shape [n_samples, n_features]
+            y (pd.Series, optional): the target training data of length [n_samples]
+
+        Returns:
+            self
+
+        Raises:
+            ValueError: if the KNNImputer receives a dataframe with both Boolean and Categorical data.
+
+        """
+        X = infer_feature_types(X)
+
+        nan_ratio = X.ww.describe().loc["nan_count"] / X.shape[0]
+        self._all_null_cols = nan_ratio[nan_ratio == 1].index.tolist()
+
+        X, _ = drop_natural_language_columns(X)
+
+        # If the Dataframe only had natural language columns, do nothing.
+        if X.shape[1] == 0:
+            return self
+
+        self._component_obj.fit(X, y)
+        return self
+
+    def transform(self, X, y=None):
+        """Transforms input by imputing missing values. 'None' and np.nan values are treated as the same.
+
+        Args:
+            X (pd.DataFrame): Data to transform.
+            y (pd.Series, optional): Ignored.
+
+        Returns:
+            pd.DataFrame: Transformed X
+        """
+        X = infer_feature_types(X)
+        not_all_null_cols = [col for col in X.columns if col not in self._all_null_cols]
+        original_index = X.index
+
+        # Drop natural language columns and transform the other columns
+        X_t, natural_language_cols = drop_natural_language_columns(X)
+        if X_t.shape[1] == 0:
+            return X
+        not_all_null_or_natural_language_cols = [
+            col for col in not_all_null_cols if col not in natural_language_cols
+        ]
+
+        X_t = self._component_obj.transform(X_t)
+        X_t = pd.DataFrame(X_t, columns=not_all_null_or_natural_language_cols)
+        X_schema = X.ww.schema
+
+        original_X_schema = X_schema.get_subset_schema(
+            subset_cols=X_schema._filter_cols(exclude=["BooleanNullable"]),
+        )
+        X_bool_nullable_cols = X_schema._filter_cols(include=["Categorical"])
+        new_ltypes_for_bool_nullable_cols = {
+            col: "Boolean" for col in X_bool_nullable_cols
+        }
```

Comment:
I rewrote the code above since honestly it confused even me reading it now. `downcast_nullable_types` turns booleans into doubles which I would prefer not to have.

Reply:
Hm, `downcast_nullable_types` appears to turn `BooleanNullable` into boolean, if you were getting doubles that seems like a problem we should address. It's not relevant to this issue though, and the cleaned up code is a lot easier to read so I'm fine with it!